core: use truthiness for block ops is_empty check

superlopuh opened the pull request:
I recently discovered that PEP8 recommends using truthiness for collection is empty check. This makes sense to me, as a lot of collections have O(n) length calculations, including strings and our recently introduced Block linked list. Here's the relevant bit:

```
For sequences, (strings, lists, tuples), use the fact that empty sequences are false:

Yes:

if not seq:
if seq:

No:

if len(seq):
if not len(seq):
```

https://pep8.org/#programming-recommendations

AntonLydike commented:
I have strong opinions against this change. I'll try to organize my points in this response:

 1. The boolean value of a Block is not necessarily well-defined. What does it mean for a block to be "false"? Is it empty? Is it not part of the IR?
 2. I feel that "falsey"-ness and type coalescence makes code a lot worse to read sometimes. This is one of those occasions. `is_empty` is easy to read, easy to understand, and short. I wouldn't want to change this for something more ambiguous.
 3. I think the pep specifically talks about strings, lists, tuples, and not all things considered "container", although I'm not 100% certain on that.
 4. I feel that your argument about the `O(n)` nature of len, while well placed, is completely inapplicable, as `is_empty` is `O(1)` anyways.

superlopuh commented:
I agree with almost all of your points, this is why I added `is_empty` in the first place. What was confusing to me is why I hadn't seen it before, Python conventions seem to sacrifice performance in the name of clarity, but not to the point of having useless O(n) calculations all over the place. That PEP8 paragraph clarified things for me, meaning that Python went for the usual "more terse and less explicit but kinda makes sense once you know about it" with the truthiness.

Now the big question is whether it's obvious to a user that a block is a sequence. And I agree with you that it probably isn't, which is why I added `BlockOps` in the first place. Now that I think about it, that's where the truthiness check should be. I'll move it now.

compor commented:
I think the confusing point was that a `Block` has operations and arguments.

The first question I asked when I saw this interface was if an empty block can have args.

Second, I'd say if iteration is clear for a container via `__iter__`, truthiness makes sense; as does `__len__`, etc.
In other words, there is only one internal collection of items to iterate over.

webmiche commented on xdsl/ir.py:
```diff
@@ -1003,6 +1007,10 @@ def __len__(self):
             result += 1
         return result
 
+    def __bool__(self) -> bool:
+        """Returns `True` if there are no operations in this block."""
+        return self.block.first_op is not None
```
To me, this is one of the Python functionalities where I personally draw the line between nice, concise syntax and absolut madness, this being on the madness side obviously. I am in favor of explicit things, in this case in particular.

Now, I know this is exactly the kind of thing that Python users would describe as "pythonic", and if we want stay with the cool kids, we should design the framework with that in mind. Therefore, I am fine with adding this (with a sligthyly disgusted look on my face), as long as you let me keep the `is_empty` property. Then, the mad lads can use their fancy, confusing Python syntax, while me and the other saints can write proper code 😉

superlopuh replied:
Lol fine I'll add it back on `Block`. After Anton's comment, I substituted `assert False` for the implementation, and found a bunch of places where falseyness of `None` was used implicitly, with your name next to the lines in question, so feel free to consider yourself a Python cool kid :)

webmiche replied:
I'm sure I only moved those around 😇

math-fehr commented:
I am personally fine with both the `if block.ops` and `if block.is_empty` option.
I don't have a strong opinion for either, so I'll let you all come to an agreement.

AntonLydike commented on docs/Toy/toy/dialects/toy.py:
```diff
@@ -212,7 +212,7 @@ def verify_(self):
         block = self.body.blocks[0]
 
-        if not block:
+        if not block.ops:
```
much better!

AntonLydike reviewed (approved):
That's much better actually!
